Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+            logger.log("""
+                engines to disable:
+                \(engineEnablements.compactMap { !$0.value ? $0.key : nil })
+                """,
+               level: .info,
+               category: .sync)
+        }
+        return engineEnablements
+    }
+
+    public class ScopedKeyError: MaybeErrorType {
+        public let description = "No key data found for scope."
+    }
+
+    public class EncryptionKeyError: MaybeErrorType {
+        public let description = "Failed to get stored key."
+    }
+
+    public class DeviceIdError: MaybeErrorType {
+        public let description = "Failed to get deviceId."
+    }
+
+    public class NoTokenServerURLError: MaybeErrorType {
+        public let description = "Failed to get token server endpoint url."
+    }
+
+    public class EngineAndKeyRetrievalError: MaybeErrorType {
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
+        let deferred = Deferred<Maybe<([EngineIdentifier], [String: String])>>()
+        var localEncryptionKeys: [String: String] = [:]
+        var rustEngines: [String] = []
+        var registeredPlaces: Bool = false
+
+        for engine in engines {
+            switch engine {
+            case "tabs":
+                self.profile.tabs.registerWithSyncManager()
+                rustEngines.append("tabs")
+            case "passwords":
+                self.profile.logins.registerWithSyncManager()
+                if let key = try? self.profile.logins.getStoredKey() {
+                    localEncryptionKeys["passwords"] = key
+                    rustEngines.append("passwords")
+                } else {
+                    self.logger.log("Login encryption key could not be retrieved for syncing",
+                                    level: .warning,
+                                    category: .sync)
+                }
+            case "bookmarks":
+                if !registeredPlaces {
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("bookmarks")
+            case "history":
+                if !registeredPlaces {
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("history")
+            default:
+                continue
+            }
+        }
+
+        deferred.fill(Maybe(success: (rustEngines, localEncryptionKeys)))
+        return deferred
+    }
+
+    fileprivate func syncRustEngines(
+        why: MozillaAppServices.SyncReason,
+        engines: [String]
+    ) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
+        let deferred = Deferred<Maybe<MozillaAppServices.SyncResult>>()
+
+        syncQueue.async { [unowned self] in
```

Comment:
Sorry, I meant to leave I comment here. I purposely left this dispatch queue usage to illustrate an alternative because it was something Nish and I discussed and I was wondering if seeing the code might help. For sure we should only be using dispatch queues *_either_* here or in the lower level `RustSyncManagerAPI` functions.

@nbhasin2 Which is better?

Reply:
Ah, so we shouldn't create custom queues unless we need it. This is the mantra I go by and is what Apple agrees too. Ref: https://developer.apple.com/documentation/dispatch/dispatchqueue

However, there can be needs like 
- Performing task in a thread (syncQueue) that only your app is aware of
- Suspending the thread (syncQueue)
- Put a label so you can associate code specifically to your thread (syncQueue)

If you are doing any of this then we can leave `syncQueue` in our code. 

**IMPT Note:** Leaving it the way it is will not cause any harm but we should at least change `com.mozilla.firefox.sync` which is the same as Profile label to  `com.mozilla.firefox.sync.rust` to debug issues related to RustSyncManager 

**Final Thoughts** Change the label and leave the queue the way it is and we can come back to it for further review on how to leverage `syncQueue` and if we can't find anything in future then we use the Global Dispatch Queue